src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java

```diff
@@ -1085,47 +1102,139 @@ private static final class SimpleStringBuilderStrategy {
         static final MethodTypeDesc PREPEND_char    = MethodTypeDesc.of(CD_long, CD_long, CD_Array_byte, CD_char, CD_String);
         static final MethodTypeDesc PREPEND_String  = MethodTypeDesc.of(CD_long, CD_long, CD_Array_byte, CD_String, CD_String);
 
+        public static final RuntimeVisibleAnnotationsAttribute STABLE_ANNOTATION = RuntimeVisibleAnnotationsAttribute.of(
+                Annotation.of(ClassDesc.ofDescriptor("Ljdk/internal/vm/annotation/Stable;")));
+
+        private static final Consumer<FieldBuilder> FINAL_STABLE_FIELD = new Consumer<FieldBuilder>() {
+            public void accept(FieldBuilder fb) {
+                fb.withFlags(ClassFile.ACC_FINAL)
+                        .with(STABLE_ANNOTATION);
+            }
+        };
+
+        static final ReferencedKeyMap<MethodType, SoftReference<MethodHandlePair>> CACHE =
+                ReferencedKeyMap.create(true, true,
+                        new Supplier<>() {
+                            @Override
+                            public Map<ReferenceKey<MethodType>, SoftReference<MethodHandlePair>> get() {
+                                return new ConcurrentHashMap<>(64);
+                            }
+                        });
+
         private SimpleStringBuilderStrategy() {
             // no instantiation
         }
 
-        private static MethodType erase(MethodType args) {
-            var bootstrapClassLoader = String.class.getClassLoader();
-            var paramTypes = args.parameterArray();
-            boolean changed = false;
-            for (int i = 0; i < paramTypes.length; ++i) {
-                if (paramTypes[i].getClassLoader() != bootstrapClassLoader) {
-                    paramTypes[i] = Object.class;
-                    changed = true;
-                }
-            }
-            return changed ? MethodType.methodType(args.returnType(), paramTypes) : args;
-        }
+        private record MethodHandlePair(MethodHandle constructor, MethodHandle concatenator) { };
 
         private static MethodHandle generate(Lookup lookup, MethodType args, String[] constants) throws Exception {
-            lookup         = MethodHandles.Lookup.IMPL_LOOKUP;
-            var erasedArgs = erase(args);
-            var className  = getClassName(String.class);
-
-            byte[] classBytes = ClassFile.of().build(ConstantUtils.binaryNameToDesc(className),
+            lookup = MethodHandles.Lookup.IMPL_LOOKUP;
```

Comment:
Maybe do a `in(String.class)`; the IMPL_LOOKUP happens to have the right host class package (java.lang for Object lookup class), but we shouldn't depend on it.